### PR TITLE
raftstore: call proposed callback in advance

### PR DESCRIPTION
**Pull request by gengliqi**
### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/11121

Problem Summary:
This PR should be merged after https://github.com/tikv/tikv/pull/11111.

Like https://github.com/tikv/tikv/pull/11111 said, the store loop is much faster than before.
The `cmd-batch-concurrent-ready-max-count` is used to enhance the effect of batch cmd and reduce CPU overhead.
The proposals will be postponed if the number of concurrent ready is greater than `cmd-batch-concurrent-ready-max-count`.
However, this method introduces a problem that the proposed callback is postponed because it is called only when proposing the cmd.
There is an optimization named `pipelined pessimistic lock` that depends on proposed callback. So this PR can reduce the duration of `pipelined pessimistic lock`.

### What is changed and how it works?

What's Changed:

This PR tries to call proposed callback in advance to solve this problem.
In order to maintain the same success rate as before, the batched proposals will be proposed if the leader is about to transfer or a admin command is about to propose.

### Related changes

- No

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
I run tipocket `pipeline` test like https://github.com/tikv/tikv/pull/8499 did.
```
2021/10/29 14:47:42 pipeline.go:354: [info] [5h59m50s] tps: 49.00 lock_err/s: 0.00 commit_err/s: 0.00 lock_err/commit_err/total: 0/0/1023440(0.00%)
```
The result shows this PR can provide the same guarantee as before, which is if there is no abnormal leader transfer, the write operation must succeed in the near future.

Side effects

- No

### Release note 

```release-note
None.
```

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
Any test code for this changing?

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -3653,6 +3721,33 @@ where
     }
 
     fn propose_raft_command(
+        &mut self,
+        msg: RaftCmdRequest,
+        cb: Callback<EK::Snapshot>,
+        diskfullopt: DiskFullOpt,
+    ) {
+        if msg.has_admin_request()
+            && self.fsm.batch_req_builder.request.is_some()
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:3730]: How can this happen?

File: components/raftstore/src/store/fsm/peer.rs [hunk 2]
```diff
@@ -637,11 +650,47 @@ where
...
+            || self.fsm.batch_req_builder.header_checked.is_some()
+            || self.fsm.peer.pending_remove
+        {
+            return;
         }
+        let cmd = self.fsm.batch_req_builder.request.take().unwrap();
+        self.fsm.batch_req_builder.header_checked = Some(false);
+        if let Ok(None) = self.pre_propose_raft_command(&cmd) {
+            let term = self.fsm.peer.term();
+            if self.fsm.peer.pending_merge_state.is_none()
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:677]: Better make a method in peer to check these conditions.

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs:3730]: It's a defense to avoid a case that a admin cmd is proposed and it can change the epoch version but there are some requests in `batch_req_builder ` and their proposed callback are called.
In most of the case, the admin request are proposed through `PeerMsg::RaftCommand` and the `propose_batch_raft_command(true)` will be called so the code here is useless.
The `RollbackMerge` cmd is proposed through calling `propose_raft_command` directly but the `header_checked` in `batch_req_builder` should be false.
In summary, the code here can not be triggered for now.

File: components/raftstore/src/store/peer.rs [hunk 3]
```diff
@@ -3541,24 +3538,24 @@ where
         Ok(Either::Left(propose_index))
     }
 
-    fn execute_transfer_leader<T>(
+    pub fn execute_transfer_leader<T>(
```
[the remark this answers is not in the document]
Comment (gengliqi) [reply components/raftstore/src/store/peer.rs:3541]: I change this logic. Leader should call `ready_to_transfer_leader` while follower should call `execute_transfer_leader `.

**Review by youjiali1995 — COMMENTED** [review]
Changes to guarantee the success rate of proposed callbacks LGTM. But codes become more obscure because the check of ongoing admin request is scattered.

I'm curious about the necessity of blocking proposals due to concurrent readies. Seems if the request batch is big enough or there is an admin request, requests will be proposed regardless of concurrent readies. Only the batch is not big enough, it will be blocked.

**Review by youjiali1995 — APPROVED** [review]

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (BusyJay) [reply components/raftstore/src/store/fsm/peer.rs:3747]: I suggest to remove the code and make `RollbackMerge` to use `propose_batch_raft_command` instead. Or just rename `propose_batch_raft_command` to `propose_raft_command`. Useless code is the source of confusion.

File: components/raftstore/src/store/fsm/peer.rs [hunk 4]
```diff
@@ -583,7 +595,6 @@ where
                             self.propose_batch_raft_command(true);
                         }
                     } else {
-                        self.propose_batch_raft_command(true);
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:586]: I don't think it's correct. For example, delete range request may be reordered to front of other puts/deletes as `propose_raft_command` may not propose any commands.

File: components/raftstore/src/store/peer.rs [hunk 5]
```diff
@@ -564,7 +564,7 @@ where
     pub max_ts_sync_status: Arc<AtomicU64>,
 
     /// Check whether this proposal can be proposed based on its epoch.
-    cmd_epoch_checker: CmdEpochChecker<EK::Snapshot>,
+    pub cmd_epoch_checker: CmdEpochChecker<EK::Snapshot>,
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:567]: Seems can be private now.

File: components/raftstore/src/store/peer.rs [hunk 6]
```diff
@@ -3541,43 +3538,12 @@ where
...
-    fn execute_transfer_leader<T>(
+    pub fn execute_transfer_leader<T>(
         &mut self,
         ctx: &mut PollContext<EK, ER, T>,
         msg: &eraftpb::Message,
         peer_disk_usage: DiskUsage,
     ) {
-        // log_term is set by original leader, represents the term last log is written
-        // in, which should be equal to the original leader's term.
-        if msg.get_log_term() != self.term() {
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:3552]: This is still necessary.

File: components/raftstore/src/store/peer.rs [hunk 7]
```diff
@@ -4005,6 +3971,19 @@ where
         }
         false
     }
+
+    /// Check if proposed callback can be called immediately.
+    pub fn check_for_proposed_cb(&mut self, cmd: &RaftCmdRequest) -> bool {
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:3976]: How about `will_likely_propose`?

File: components/raftstore/src/store/fsm/peer.rs [hunk 8]
```diff
@@ -3652,7 +3736,25 @@ where
         }
     }
 
+    /// Propose batched raft commands(if any) first, then propose the given raft command.
     fn propose_raft_command(
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:3740]: Seems the method is unnecessary at all. The only place it's used is https://github.com/tikv/tikv/pull/11147/files#diff-7a4718b9cddd27a4ef7e563024d5283cf14b4ba3abf26edc33b974b87680a5fbL587, which can be prepend batch propose call.

Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs:3740]: Another place is `RollbackMerge`. This method is intended to be called when the order between this command and batched normal commands is necessary.

File: components/raftstore/src/store/peer.rs (hunk 6 above)
Comment (gengliqi) [reply components/raftstore/src/store/peer.rs:3552]: This check is moved to `on_transfer_leader_msg`.

Comment (BusyJay) [reply components/raftstore/src/store/peer.rs:3552]: I mean it's also necessary for follower.

Comment (gengliqi) [reply components/raftstore/src/store/peer.rs:3552]: Yes. Follower should also call `on_transfer_leader_msg` when receiving `TransferLeader` msg.

File: components/raftstore/src/store/peer.rs [hunk 9]
```diff
@@ -267,7 +267,7 @@ impl<S: Snapshot> ProposedAdminCmd<S> {
     }
 }
 
-struct CmdEpochChecker<S: Snapshot> {
+pub struct CmdEpochChecker<S: Snapshot> {
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:270]: Can be private now.

**Review by BusyJay — APPROVED** [review]
Rest LGTM
